signer/signer.go

```diff
@@ -0,0 +1,483 @@
+			return pkix.Extension{}, err
+		}
+		list.SCTList = append(list.SCTList, ctx509.SerializedSCT{Val: sctBytes})
+	}
+	listBytes, err := cttls.Marshal(list)
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+	extBytes, err := asn1.Marshal(listBytes)
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+	return pkix.Extension{
+		Id:    sctListOID,
+		Value: extBytes,
+	}, nil
+}
+
+var mustStapleExt = pkix.Extension{
+	Id:    asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 24},
+	Value: []byte{0x30, 0x03, 0x02, 0x01, 0x05},
+}
+
+func generateSKID(pk crypto.PublicKey) ([]byte, error) {
+	pkBytes, err := x509.MarshalPKIXPublicKey(pk)
+	if err != nil {
+		return nil, err
+	}
+	var pkixPublicKey struct {
+		Algo      pkix.AlgorithmIdentifier
+		BitString asn1.BitString
+	}
+	if _, err := asn1.Unmarshal(pkBytes, &pkixPublicKey); err != nil {
+		return nil, err
+	}
+	skid := sha1.Sum(pkixPublicKey.BitString.Bytes)
+	return skid[:], nil
+}
+
+// Issue generates a certificate from the provided issuance request and
+// signs it. Before the signing the certificate with the issuers private
+// key it is signed using a throwaway key so that it can be linted using
+// zlint. If the linting fails an error is returned and the certificate
+// is not signed using the issuers key.
+func (s *Signer) Issue(req *IssuanceRequest) ([]byte, error) {
+	// check request is valid according to the issuance profile
+	if err := s.profile.requestValid(s.clk, req); err != nil {
+		return nil, err
+	}
+
+	// generate template from the issuance profile
+	template := s.profile.generateTemplate(s.clk)
+
+	// populate template from the issuance request
+	template.PublicKey = req.PublicKey
+	template.NotBefore, template.NotAfter = req.NotBefore, req.NotAfter
+	template.SerialNumber = big.NewInt(0).SetBytes(req.Serial)
+	if req.CommonName != "" {
+		template.Subject.CommonName = req.CommonName
+	}
+	template.DNSNames = req.DNSNames
+	template.AuthorityKeyId = s.issuer.SubjectKeyId
+	skid, err := generateSKID(req.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	template.SubjectKeyId = skid
+	switch template.PublicKey.(type) {
+	case *rsa.PublicKey:
+		template.KeyUsage = x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment
+	case *ecdsa.PublicKey:
+		template.KeyUsage = x509.KeyUsageDigitalSignature
+	}
+
+	if req.IncludeCTPoison {
+		template.ExtraExtensions = append(template.ExtraExtensions, ctPoisonExt)
+	} else if req.SCTList != nil {
+		sctListExt, err := generateSCTListExt(req.SCTList)
+		if err != nil {
+			return nil, err
+		}
+		template.ExtraExtensions = append(template.ExtraExtensions, sctListExt)
+	}
+
+	if req.IncludeMustStaple {
+		template.ExtraExtensions = append(template.ExtraExtensions, mustStapleExt)
+	}
+
+	// check that the tbsCertificate is properly formed by signing it
+	// with a throwaway key and then linting it using zlint
+	lintCertBytes, err := x509.CreateCertificate(rand.Reader, template, s.issuer, template.PublicKey, s.lintKey)
```

Comment:
Line 390 sets `template.PublicKey = req.PublicKey`, so be consistent in which one is used here vs. on line 445.